aten/src/ATen/cuda/CUDAEvent.h

```diff
@@ -17,19 +17,36 @@ namespace at { namespace cuda {
 /*
 * CUDAEvents are movable not copyable wrappers around CUDA's events.
 *
-* CUDAEvents are constructed lazily when recorded on streams. The events
-* have a device, and this device is acquired from the first recording stream.
-* Later streams that record to the event must share this device, but streams
-* on any device can wait on the event.
+* CUDAEvents are constructed lazily when recorded unless it is reconstructed
+* from a cudaIpcEventHandle_t. The events have a device, and this device is
+* acquired from the first recording stream or passed in when reconstructed from
+* a handle. Later streams that record to the event must share this device, but
+* streams on any device can wait on the event.
 */
 struct AT_CUDA_API CUDAEvent {
   // Constants
   static constexpr unsigned int DEFAULT_FLAGS = cudaEventDisableTiming;
 
   // Constructors
-  CUDAEvent(unsigned int flags = DEFAULT_FLAGS)
+  explicit CUDAEvent(unsigned int flags = DEFAULT_FLAGS)
   : flags_{flags} { }
 
+  // Note: the original event and the reconstructed event now share recorded
+  // activities. Users need to make sure the last recording event (either
+  // original or reconstructed) must not be destructed when synchronize() is
+  // called. Otherwise, the behavior is undefined.
```

Comment:
I'm confused by the description here. I don't see what's special about synchronize(). In general, it's undefined behavior to call any function on an imported event after the exported event has been freed.

Perhaps it's better to paraphrase the CUDA documentation. Something like:

```
Opens an interprocess event handle exported from another process.

The imported event behaves like a locally created event with the cudaEventDisableTiming flag
specified. Performing operations on the imported event after the exported event has been freed
will result in undefined behavior.
```

Reply:
Sorry, I forgot to clean up this doc before that push. I updated the doc in subsequent push. I think it is a little bit more complicated than CUDA's official doc. For example, the official doc did not state whether we can all methods on the exported event when the imported event is destructed. From what I see, I think the dependency is symmetric. Whenever a method (except record?) is called, that last recording event (can be either exported or imported event) must not be destructed.

Reply:
> For example, the official doc did not state whether we can all methods on the exported event when the imported event is destructed.

This can be assumed and need not be stated.

> From what I see, I think the dependency is symmetric. Whenever a method (except record?) is called, that last recording event (can be either exported or imported event) must not be destructed.

I checked and I don't think this is the case. You can call synchronize(), wait(), and query() on the original event even if it's only recorded (and deleted) in another process.

Reply:
Yes, I just verified, you are right. I might have misread some test results yesterday.